rfc/0003-indexmanagement.md

```diff
@@ -0,0 +1,156 @@
+
+```java
+String name
+boolean isPrimary
+IndexType type //enum type GSI/VIEW
+String rawType //raw string type in case a new index type is introduced, for forward-compatibility
+String state //eg. "online", "pending"
+String keyspace
+String namespace
+JsonArray indexKey //or equivalent Json representation
+```
+
+## Creating Indexes
+There are two specificities for index creation (both of primary and secondary indexes): the index could already exist and there is the possibility of deferring the build.
+
+In the first specificity, it makes sense to offer two alternatives: either this is considered an error (as N1QL does) or the creation should be purely and simply ignored.
+This choice can be made through a first `boolean` parameter, `ignoreIfExist`. If it is `false`, then a SDK-related exception should be raised (eg. in Java a `CouchbaseException`).
+
+The second specificity can be triggered by another `boolean` parameter, `defer`. If set to true, the N1QL query will use the "with defer" syntax and the index will simply be "pending".
+
+Signatures:
+```java
+boolean createPrimaryIndex(boolean ignoreIfExist, boolean defer)
+
+boolean createIndex(String indexName, List<String> fields, boolean ignoreIfExist, boolean defer)
+```
+
+## Dropping Indexes
+For index removal, the inverse of `ignoreIfExist` must be put in place: N1QL considers it an error to attempt dropping an index that doesn't exist. So the `ignoreIfNotExist` will drive the behavior of the API in such a case: either `false` and a `CouchbaseException` (or equivalent) will be thrown, or `true` and the invocation will do nothing.
+
+Signatures:
+```java
+boolean dropPrimaryIndex(boolean ignoreIfNotExist)
+
+boolean dropIndex(String name, boolean ignoreIfNotExist)
+```
+
+> **Note**: Unlike index creation, defer the drop of an index is not possible
+
+## Building deferred indexes
+Signature:
+```java
+List<String> buildDeferredIndexes()
+```
+
+Remarks:
+The returned list is the list of indexes that were in **`pending`** state at the time of the call.
+
+The method will first list indexes, only keeping the `pending` ones. It then issues the appropriate N1QL statement from the list of pending indexes (here `INDEX_NAME1` and `INDEX_NAME3`):
+
+```sql
+BUILD INDEX ON `BUCKET_NAME_INJECTED_HERE`(`INDEX_NAME1`, `INDEX_NAME3`)
+```
+
+# Watching Indexes
+>**Note**: this part of the API could be considered optional.
+
+Since the process of building a deferred index is asynchronous, the server immediately returns control to the client. It would be useful to have an API to poll the list of indexes until one (or several) indexes are seen as "online". In the blocking API, this would allow to make sure that an index is ready.
+
+Signature:
+```java
+boolean watchIndex(String indexName, long watchTimeout, TimeUnit watchTimeUnit)
+
+List<IndexInfo> watchIndexes(List<String> watchList, boolean watchPrimary, long watchTimeout, TimeUnit watchTimeUnit)
+```
```

Comment:
I am wondering if a callback/event driven API might be better here (or as an option) rather than a blocking API might fit this better? I can see in some cases where I would want to block and in others (very log index build times), i wouldn't...I would like to handle an event and then continue code exec from there.